### ISSUE TITLE
Mega Issue: Karpenter doesnt support custom resources requests/limit

### ISSUE THREAD

**Issue by prateekkhera** [aws/karpenter-core#751]
## Version

**Karpenter**: v0.10.1

**Kubernetes**: v1.20.15

## Expected Behavior
Karpenter should be able to trigger an autoscale

## Actual Behavior
Karpenter isnt able to trigger an autoscale

## Steps to Reproduce the Problem
We're using Karpenter on EKS. We have pods that has custom resource requests/limits in their spec definition - `smarter-devices/fuse: 1`. Karpenter seems to not respecting this resource and fails to autoscale and the pod remains to be in pending state 

## Resource Specs and Logs

Provisioner spec

```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  limits:
    resources:
      cpu: "100"
  provider:
    launchTemplate: xxxxx
    subnetSelector:
      xxxxx: xxxxx
  requirements:
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - on-demand
  - key: node.kubernetes.io/instance-type
    operator: In
    values:
    - m5.large
    - m5.2xlarge
    - m5.4xlarge
    - m5.8xlarge
    - m5.12xlarge
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  ttlSecondsAfterEmpty: 30
status:
  resources:
    cpu: "32"
    memory: 128830948Ki
```

pod spec

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: fuse-test
  labels:
    app: fuse-test
spec:
  replicas: 1
  selector:
    matchLabels:
      name: fuse-test
  template:
    metadata:
      labels:
        name: fuse-test
    spec:
      containers:
      - name: fuse-test
        image: ubuntu:latest
        ports:
          - containerPort: 8080
            name: web
            protocol: TCP
        securityContext:
          capabilities:
            add:
              - SYS_ADMIN
        resources:
          limits:
            cpu: 32
            memory: 4Gi
            smarter-devices/fuse: 1  # Custom resource
          requests:
            cpu: 32
            memory: 2Gi
            smarter-devices/fuse: 1  # Custom resource
        env:
        - name: S3_BUCKET
          value: test-s3
        - name: S3_REGION
          value: eu-west-1
```

karpenter controller logs:

controller 2022-06-06T15:59:00.499Z    ERROR    controller    no instance type satisfied resources {"cpu":"32","memory":"2Gi","pods":"1","smarter-devices/fuse":"1"} and requirements kubernetes.io/os In [linux], karpenter.sh/capacity-type In [on-demand], kubernetes.io/hostname In [hostname-placeholder-3403], node.kubernetes.io/instance-type In [m5.12xlarge m5.2xlarge m5.4xlarge m5.8xlarge m5.large], karpenter.sh/provisioner-name In [default], topology.kubernetes.io/zone In [eu-west-1a eu-west-1b], kubernetes.io/arch In [amd64];

**Comment by project-administrator** [comment]
We tried enabling the **hugepages** on all nodes with `"vm.nr_hugepages" = "2048"`.

After this karpenter went crazy spinning up 50 new worker nodes for one of the existing pods. That pod does not have anything related to **hugepages**, just a RAM requests of 3Gb. Karpenter spins up a new node with 8Gb of RAM and after scheduler is not able to run the pod on the new node. Then, karpenter spins up the new node (again, with 8Gb of RAM), and again scheduler can't run the pod. 

It looks like **hugepages** option messes up the karpenter ability to calculate the memory resources properly.

**Comment by chomatdam** [comment]
We're facing the same issue with KubeVirt. Given that it's been ongoing for a while, it might be good to consider both a short-term solution to unblock and a long-term solution?

I noticed PR https://github.com/kubernetes-sigs/karpenter/pull/603 mentioning the deprecated Karpenter config map. As an alternative, I created a fork using the same approach but sourcing configuration from options (arg or environment variable).
Would this be an interesting direction to explore? Or is the current state of this issue more "not a priority, maintain your forks until we have a better design / long-term approach for it"?

**Comment by garvinp-stripe** [comment]
Bringing in some context on huge pages which I think is more problematic that just "defining custom allocable". Huge pages are essentially a user configurable based on a mix of instance type + user's need. That means that you could have different huge page allocable even within the same instance type based on what the node is used for. To add to this problem, hugepages are pre-allocated at boot time set at the linux level so it at best can be set in NodeClass level and must be passed through via startup script of the node BUT because nodeclasses can be used by different instance types nodeclass itself cannot be relied upon to know ahead of time how much hugepage resource can be provided.  

What does this all mean? Implementation would be difficult because if we want Karpenter to work Karpenter would need to know ahead of time a mapping of all permutation of instance type + possible hugepages. This means user must input a mapping in node pool for instance types to huge pages resource in addition to instance type to huge page in nodeclass to specify how instances come up. 

I am likely missing a few pieces of this puzzle but this what I think needs to be solved for hugepages

**Comment by jonathan-innis** [comment]
> I am likely missing a few pieces of this puzzle but this what I think needs to be solved for hugepages

I think there are a couple simplifications that we could do here to support hugepages if we wanted to:
1. Consider the entire available memory to be used for hugepages. Add up all of the hugepages into the resource requests for the NodeClaim and then launch an instance, configuring the startup script to start with that many hugepages so that all of the pods can schedule
2. Allow users to configure a percentage of the memory to be allocated to hugepages. We would calculate the hugepages during the GetInstanceTypes() call and then use that for scheduling. If we allowed this, this would most likely be a setting on the NodeClass and then we would just pass it down through the GetInstanceTypes() call from the CloudProvider.

**Comment by garvinp-stripe** [comment]
> Consider the entire available memory to be used for hugepages. Add up all of the hugepages into the resource requests for the NodeClaim and then launch an instance, configuring the startup script to start with that many hugepages so that all of the pods can schedule

One issue here is huge pages are carved out of memory, for us this doesn't matter because we actually do want to move to all huge pages but most users likely have a mix of huge pages and normal memory. If you advertise all huge pages then your nodes technically doesn't have memory

> Allow users to configure a percentage of the memory to be allocated to hugepages. We would calculate the hugepages during the GetInstanceTypes() call and then use that for scheduling. If we allowed this, this would most likely be a setting on the NodeClass and then we would just pass it down through the GetInstanceTypes() call from the CloudProvider.

That feels reasonable and removes the need to map instance types to huge pages. Once again that works for us but I am unsure if other users have more unique configuration

**Comment by garvinp-stripe** [comment]
> Yeah, it's a little tough to boil the ocean here without creating wayyyyy too much configuration and making this likely unreasonable to manage for users.

Agreed. I think this general approach should work, need some time to bake in my head if we would encounter any issues.

**Comment by uniemimu** [comment]
We also need this feature. Our use-case is related to a controller which creates extended resources to nodes immediately when a new node is created. Karpenter will not create a node for pods using such extended resources, because it doesn't understand the extended resources.

In our case, using node affinity and node selectors together with existing node labels is sufficient to direct Karpenter to pick a good node. The only thing we need is Karpenter to ignore a list of extended resources, when finding the correct instance type. Having said that, I do have a forked workaround, but forked workarounds are not acceptable where I work, for good reason.

Having ignorable extended resources wouldn't be new in Kubernetes. They exist also in the [scheduler](https://kubernetes.io/docs/reference/config-api/kube-scheduler-config.v1/#kubescheduler-config-k8s-io-v1-ExtenderManagedResource).

**Comment by garvinp-stripe** [comment]
How much appetite is there to simply have an override config map that has per instance type override on resources capacity just for Karpenter simulation (Support for huge pages and possibly other extend resources!?!)? 
https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instancetype/types.go#L179

Config map that as instance types + any resource override and if a particular resource isn't override, take what is provided from cloud provider. 

Pin the configmap per NodeClass via a new setting on NodeClass `instanceTypeResourceOverride`. Note that changes to the configmap won't be reflected on current nodes, we would use drift to reconcile the changes.

**Comment by johngmyers** [comment]
Hopefully users wouldn't need to maintain their own list of acceptable instance types in order to handle the "fuse" use case, as fuse doesn't depend on particular instance types.

It's a bit frustrating that the fuse use case is being held up by hugepages. The fuse use case is probably common enough to justify being handled out of the box.

**Comment by GnatorX** [comment]
I think fuse's use case is not the same as hugepages and shouldn't be tied together. Fuse likely can do https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/ ?

**Comment by GnatorX** [comment]
> A nice bonus would be if one could also define extended resources which are applied to all instance types, covering in a simple fashion the fuse-case.

That feels fine also. Let me try to bring this up during working group meeting

**Comment by Bourne-ID** [comment]
Curious if this could be a configuration on the NodePool; we're able to add custom Requirements to allow Karpenter to schedule when hard affinities or tolerances are defined. Would having an entry to define _capacity hints_ to karpenter 'this node pool will satisfy requests/limits for [custom capacity]' be an option? 

My use case is smarter-devices/kvm - which can be filtered on a nodepool as metal. I could imagine the same for hugePages or similar - we know what the instances which has these so we can filter them using custom NodePools. 

By using weighting we can define these after the main nodepools - so in my example, I would have Spot for all weight 100, on-demand for all weight 90, and then our KVM with capacity hints of 80. 

In the mean time, I'm using an Overprovisioner marked with hard affinity for metal instances to ensure these pods can be scheduled; it's a tradeoff with extra cost but the ability to use Karpenter exclusively.

**Comment by ellistarn** [comment]
I wonder if this is something that might be useful to configure both at the node pool level, and at the instance type level. Ultimately, we were learning away from an InstanceTypeOverride CRD due to the level of effort to configure it, but perhaps with support for both, it provides an escape hatch as well as the ability to define a simple blanket policy.

We could choose any/all of the following:

1. Cloudprovider automatically knows extended resource values (e.g. GPU) 
2. NodePool (or class) lets you specify a flat resource value per node pool
3. NodePool (or class) lets you specify a scalar resource values (e.g. hugePageMemoryPercent) 
4. InstanceType CRD (or config map) let's you define per instance type resource overrides.

**Comment by Bryce-Soghigian** [comment]
https://docs.google.com/document/d/1vEdd226PYlGmJqs6gWlC2pTyDKhZE8DyCU2SbNB35wM/edit Looking for user stories from customers on their extended resources support story here. Please leave some comments here!

After we feel confident we have captured all the critical usecases, I will go through and propose some RFCs to solve the various dimensions of these problems.

**Comment by AaronFriel** [comment]
@jmickey has captured my comments above in the doc.

**Comment by ellistarn** [comment]
For anyone watching this issue, I have a proof of concept to solve this problem here: https://github.com/kubernetes-sigs/karpenter/pull/1305

**Comment by daverin** [comment]
Searching for a fix. My current workaround is absolutely hideous:

```yaml
         resources:
            requests:
              cpu: 4000m
              memory: 16Gi
            limits:
              cpu: 11000m
              memory: 24Gi
              xilinx.com/fpga-xilinx_u30_gen3x4_base_2-0: 1
              # ---
              # karpenter will not provision this node if this custom device is here. 
              # To provision, comment out, wait for launch and daemon set to launch
              # then uncomment, sync kustomization, then kill the old pod
```

Does anyone have a better workaround?

**Comment by poussa** [comment]
I guess feature request did not make the v1.0 release. Can someone confirm?

**Comment by sergeykranga** [comment]
Hi everyone,

We too faced this exact issue. Karpenter consolidation stopped working after a third party software started to add extended resources to our pods and nodes (we run Spark on EKS). This increased costs significantly and we couldn't wait until this PR is merged/released.

I would like to propose a temporary solution to those who, just like us, need a quick workaround and, **importantly, don't care about the extended resources being part of Karpenter's scheduling simulation to find replacement nodes**. If you are looking for the inverse and want Karpenter to take extended resources into account, this will not work for you.

This workaround builds a custom Karpenter controller image. I would like to thank @jonathan-innis for his [PR](https://github.com/kubernetes-sigs/karpenter/pull/603), where the required code changes come from.

It is assumed that you can publish the image to your own repository and use it in your Karpenter deployment (in our environment we use ECR and EKS, Karpenter is deployed with Helm using the official chart).

The steps:
1. Fork the https://github.com/kubernetes-sigs/karpenter repo and clone it
2. To ensure Karpenter version consistency, from the [Releases](https://github.com/kubernetes-sigs/karpenter/releases) page, select the commit of the version you are interested in (in our case we use [0.37.0](https://github.com/kubernetes-sigs/karpenter/releases/tag/v0.37.0)) so the commit is 38b4c32043724f6dafce910669777877295d89ec). Checkout this commit locally, you'll get into detached head state
3. Create a branch for the fix (i.e. `ignore_extended_resources`)
4. Make the same changes as shown in the original [pull request](https://github.com/kubernetes-sigs/karpenter/pull/603). **Make sure to change** `smarter-devices` to your own extended resource name that you would like to ignore
5. Commit and push the branch to your fork
6. Clone the https://github.com/aws/karpenter-provider-aws repo (no need to fork)
7. Do the step 2 for this repo too, to ensure the provider and karpenter versions match
8. Install [`ko`](https://ko.build/install/) and `go`
9. Add this line to the end of `go.mod` file (change the branch name if it is different):
```
replace sigs.k8s.io/karpenter => github.com/<your-org-or-account-name>/karpenter ignore_extended_resources
```
10. Run `go mod tidy` from the root of the repo, it should download the replacement dependency from your fork
```
$ go mod tidy
go: downloading github.com/<your-org-or-account-name>/karpenter v0.0.0-20241003171933-4834e30c1e45
go: downloading k8s.io/cloud-provider v0.30.1
go: downloading k8s.io/component-base v0.30.1
go: downloading k8s.io/csi-translation-lib v0.30.1
go: downloading github.com/stretchr/testify v1.9.0
go: downloading go.uber.org/goleak v1.3.0
go: downloading gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c
go: downloading go.uber.org/automaxprocs v1.5.3
go: downloading github.com/jmespath/go-jmespath/internal/testify v1.5.1
go: downloading github.com/blang/semver/v4 v4.0.0
go: downloading golang.org/x/mod v0.17.0
go: downloading github.com/kr/pretty v0.3.1
go: downloading github.com/spf13/cobra v1.8.0
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading github.com/golang/glog v1.1.0
go: downloading github.com/prashantv/gostub v1.1.0
go: downloading github.com/inconshreveable/mousetrap v1.1.0
go: downloading github.com/rogpeppe/go-internal v1.10.0
go: downloading github.com/kr/text v0.2.0
```
11. We didn't want the build process to publish the image anywhere yet, so we modified the `Makefile` to only build it locally. If you want the same behavior, modify the `Makefile` like this:
```
-       $(eval CONTROLLER_IMG=$(shell $(WITH_GOFLAGS) KOCACHE=$(KOCACHE) KO_DOCKER_REPO="$(KO_DOCKER_REPO)" ko build --bare github.com/aws/karpenter-provider-aws/cmd/controller))
+       $(eval CONTROLLER_IMG=$(shell $(WITH_GOFLAGS) ko build --bare -L github.com/aws/karpenter-provider-aws/cmd/controller))
```
12. Ensure Docker is running on your machine
13. Build the image:
```
$ make image
2024/10/03 20:24:38 Using base public.ecr.aws/eks-distro-build-tooling/eks-distro-minimal-base@sha256:5c9f8d3da61beb5ed4d60344767c7efaf24e76f6e1355a86697b507b5488a4d6 for github.com/aws/karpenter-provider-aws/cmd/controller
2024/10/03 20:24:39 Building github.com/aws/karpenter-provider-aws/cmd/controller for linux/amd64
2024/10/03 20:24:44 Loading ko.local:63c03b831c126d0b62d9df69a4d0cfbda8a57092bf7f2db08db8719ec05f20fa
2024/10/03 20:24:46 Loaded ko.local:63c03b831c126d0b62d9df69a4d0cfbda8a57092bf7f2db08db8719ec05f20fa
2024/10/03 20:24:46 Adding tag latest
2024/10/03 20:24:46 Added tag latest
make: Nothing to be done for 'image'.
```
14. The `ko.local` is the karpenter controller image.

At this point, you can upload it to the repository of your choice, update the `controller.image.repository`, `controller.image.tag` and `controller.image.digest` Helm chart variables and deploy it.

We stopped seeing any `could not schedule pod` errors from Karpenter instantly, and consolidation started working properly. Thanks again @jonathan-innis !

P.S.
If you run YuniKorn in a cloud environment, don't forget to update the [node sorting policy](https://yunikorn.apache.org/docs/next/user_guide/sorting_policies/#binpackingpolicy) to `binpacking` to achieve better node utilization and reduce costs!

**Comment by gillg** [comment]
> For anyone watching this issue, I have a proof of concept to solve this problem here: #1305

This seems to be a universal way to go. Probably a bit complex to configure and document the first time, but it extends a lot the capabilities. Any plans ready for an implementation ?

@ellistarn to simplify a little bit don't you think it would make sense to be able using a selector pointing directly to a NodePool name ? Maybe I miss some logic in NodeOverlays, but as an example some resources like `xilinx.com/fpga-xilinx_u30_gen3x4_base_2-0: 1` require special hardware based on a [Device plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/) (question could be continued in the PR directly)

**Comment by mikeage** [comment]
/remove-lifecycle stale

This would be incredibly helpful for us.

**Comment by nick123pig** [comment]
While we all anxiously await NodeOverlay, I found another workaround. On v1.3.3 of AWS Karpenter, I found you can patch the computeCapacity method to add whatever resources you need. For example, the below patch will allow the use of karpenter with kubevirt.

```
diff --git a/pkg/providers/instancetype/types.go b/pkg/providers/instancetype/types.go
index 7e917898..5bfb5660 100644
--- a/pkg/providers/instancetype/types.go
+++ b/pkg/providers/instancetype/types.go
@@ -319,6 +319,13 @@ func computeCapacity(ctx context.Context, info ec2types.InstanceTypeInfo, amiFam
                v1.ResourceHabanaGaudi:          *habanaGaudis(info),
                v1.ResourceEFA:                  *efas(info),
        }
+
+       if strings.Contains(string(info.InstanceType), ".metal") {
+               resourceList["devices.kubevirt.io/kvm"] = resource.MustParse("1")
+               resourceList["devices.kubevirt.io/tun"] = resource.MustParse("1")
+               resourceList["devices.kubevirt.io/vhost-net"] = resource.MustParse("1")
+       }
+
        return resourceList
 }
```

**Comment by sergiy-slobodyan** [comment]
Node Overlay addresses this issue.
https://karpenter.sh/docs/concepts/nodeoverlays/
